core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+      throw new IllegalStateException(s"Preparing transaction state transition to $newState " +
+        s"while it already a pending state ${pendingState.get}")
 
     // check that the new state transition is valid and update the pending state if necessary
     if (TransactionMetadata.validPreviousStates(newState).contains(state)) {
       pendingState = Some(newState)
-      true
+
+      TransactionMetadataTransition(producerId, newEpoch, newTxnTimeoutMs, newState, newTopicPartitions, newTxnStartTimestamp, updateTimestamp)
     } else {
-      false
+      throw new IllegalStateException(s"Preparing transaction state transition to $newState failed since the target state" +
+        s" $newState is not a valid previous state of the current state $state")
     }
   }
 
-  def completeTransitionTo(newState: TransactionState): Boolean = {
+  def completeTransitionTo(newMetadata: TransactionMetadataTransition): Boolean = {
+    // metadata transition is valid only if all the following conditions are met:
+    //
+    // 1. the new state is already indicated in the pending state.
+    // 2. the pid is the same (i.e. this field should never be changed)
+    // 3. the epoch should be either the same value or old value + 1.
+    // 4. the last update time is no smaller than the old value.
+    // 4. the old partitions set is a subset of the new partitions set.
+    //
+    // plus, we should only try to update the metadata after the corresponding log entry has been successfully written and replicated (see TransactionStateManager#appendTransactionToLog)
+    //
+    // if valid, transition is done via overwriting the whole object to ensure synchronization
+
     val toState = pendingState.getOrElse(throw new IllegalStateException("Completing transaction state transition while it does not have a pending state"))
-    if (toState != newState) {
+
+    if (toState != newMetadata.txnState ||
+      producerId != newMetadata.producerId ||
+      txnLastUpdateTimestamp <= newMetadata.txnLastUpdateTimestamp) {
+
       false
     } else {
-      pendingState = None
-      state = toState
-      true
+      val updated = toState match {
+        case Empty => // from initPid
+          if (producerEpoch > newMetadata.producerEpoch ||
+            producerEpoch < newMetadata.producerEpoch - 1 ||
+            newMetadata.topicPartitions.nonEmpty ||
+            newMetadata.txnStartTimestamp != -1) {
+
+            false
+          } else {
+            txnTimeoutMs = newMetadata.txnTimeoutMs
+            producerEpoch = newMetadata.producerEpoch
+
+            true
+          }
+
+        case Ongoing => // from addPartitions
+          if (producerEpoch != newMetadata.producerEpoch ||
+            !topicPartitions.subsetOf(newMetadata.topicPartitions) ||
+            txnTimeoutMs != newMetadata.txnTimeoutMs ||
+            txnStartTimestamp < newMetadata.txnStartTimestamp) {
```

Comment:
That is true: when the pid is first created, there is no txn yet, the startTime is set to "-1", when we received the first addPartitions we will update the startTime to `now`, and it will not be updated until it has completed, and then a new addPartition is received indicating the start of a new txn for this pid.